apiserver/facades/agent/uniter/uniter.go

```diff
@@ -1776,82 +1751,63 @@ func (u *UniterAPI) WatchRelationUnits(ctx context.Context, args params.Relation
 // SetRelationStatus updates the status of the specified relations.
 func (u *UniterAPI) SetRelationStatus(ctx context.Context, args params.RelationStatusArgs) (params.ErrorResults, error) {
 	var statusResults params.ErrorResults
-
-	unitCache := make(map[string]*state.Unit)
-	getUnit := func(tag string) (*state.Unit, error) {
-		if unit, ok := unitCache[tag]; ok {
-			return unit, nil
-		}
-		unitTag, err := names.ParseUnitTag(tag)
+	results := make([]params.ErrorResult, len(args.Args))
+	for i, arg := range args.Args {
+		unitTag, err := names.ParseUnitTag(arg.UnitTag)
 		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		unit, err := u.st.Unit(unitTag.Id())
-		if errors.Is(err, errors.NotFound) {
-			return nil, apiservererrors.ErrPerm
+			results[i].Error = apiservererrors.ServerError(err)
+			continue
 		}
+		unitName, err := coreunit.NewName(unitTag.Id())
 		if err != nil {
-			return nil, errors.Trace(err)
+			results[i].Error = apiservererrors.ServerError(err)
+			continue
 		}
-		unitCache[tag] = unit
-		return unit, nil
+		err = u.oneSetRelationStatus(ctx, unitName, arg.RelationId,
+			arg.Status, arg.Message)
+		results[i].Error = apiservererrors.ServerError(err)
 	}
+	statusResults.Results = results
+	return statusResults, nil
+}
 
-	checker := u.leadershipChecker
-	changeOne := func(arg params.RelationStatusArg) error {
-		// TODO(wallyworld) - the token should be passed to SetStatus() but the
-		// interface method doesn't allow for that yet.
-		unitTag := arg.UnitTag
-		if unitTag == "" {
-			// Older clients don't pass in the unit tag explicitly.
-			unitTag = u.auth.GetAuthTag().String()
-		}
-		unit, err := getUnit(unitTag)
-		if err != nil {
-			return err
-		}
-		token := checker.LeadershipCheck(unit.ApplicationName(), unit.Name())
-		if err := token.Check(); err != nil {
-			return errors.Trace(err)
-		}
+func (u *UniterAPI) oneSetRelationStatus(
+	ctx context.Context,
+	unitName coreunit.Name,
+	relID int,
+	relStatus params.RelationStatusValue,
+	message string,
+) error {
+	// Verify the relation exist before continuing.
+	relationUUID, err := u.relationService.GetRelationUUIDByID(ctx, relID)
+	if errors.Is(err, errors.NotFound) {
+		return apiservererrors.ErrPerm
```

Comment:
This is the wrong error, we should output the domain error.

Reply:
We do not have relation domain error types at this time. The facade changes are being made first. We can be more specific at that time. However differentiating between NotFounds isn't interesting in the facade.

The current facade implementation says that any NotFound returns an ErrPerm. I haven't changed that here.

Reply:
> We do not have relation domain error types at this time.

You should add a comment/task to ensure you don't forget to update them.

> The current facade implementation says that any NotFound returns an ErrPerm.

That's fine, I'm not asking for that :)